hack/verify-staticcheck.sh

```diff
@@ -0,0 +1,154 @@
+IGNORE=(
+  "vendor/k8s.io/kubectl/pkg/cmd/edit/testdata"  # golang/go#24854, dominikh/go-tools#565
+)
+export IFS='|'; ignore_pattern="^(${IGNORE[*]})\$"; unset IFS
+
+# Ensure that we find the binaries we build before anything else.
+export GOBIN="${KUBE_OUTPUT_BINPATH}"
+PATH="${GOBIN}:${PATH}"
+
+# Install staticcheck from vendor
+echo 'installing staticcheck from vendor'
+go install k8s.io/kubernetes/vendor/honnef.co/go/tools/cmd/staticcheck
+
+cd "${KUBE_ROOT}"
+
+# Check that the file is in alphabetical order
+failure_file="${KUBE_ROOT}/hack/.staticcheck_failures"
+kube::util::check-file-in-alphabetical-order "${failure_file}"
+
+export IFS=$'\n'
+# NOTE: when "go list -e ./..." is run within GOPATH, it turns the k8s.io/kubernetes
+# as the prefix, however if we run it outside it returns the full path of the file
+# with a leading underscore. We'll need to support both scenarios for all_packages.
+all_packages=()
+while IFS='' read -r line; do
+  all_packages+=("./$line")
+done < <( hack/make-rules/helpers/cache_go_dirs.sh "${KUBE_ROOT}/_tmp/all_go_dirs" |
+            grep "^${FOCUS:-.}" |
+            grep -vE "(third_party|generated|clientset_generated|/_)" |
+            grep -vE "$ignore_pattern" )
+
+failing_packages=()
+if [[ -z $FOCUS ]]; then # Ignore failing_packages in FOCUS mode
+  while IFS='' read -r line; do failing_packages+=("$line"); done < <(cat "$failure_file")
+fi
+unset IFS
+errors=()
+not_failing=()
+
+while read -r error; do
+  # Ignore compile errors caused by lack of files due to build tags.
+  # TODO: Add verification for these directories.
+  ignore_no_files="^-: build constraints exclude all Go files in .* \(compile\)"
+  if [[ $error =~ $ignore_no_files ]]; then
+    continue
+  fi
+
+  file="${error%%:*}"
+  pkg="$(dirname "$file")"
+  kube::util::array_contains "$pkg" "${failing_packages[@]}" && in_failing=$? || in_failing=$?
+  if [[ "${in_failing}" -ne "0" ]]; then
+    errors+=( "${error}" )
+  elif [[ "${in_failing}" -eq "0" ]]; then
+    really_failing+=( "$pkg" )
+  fi
+done < <(staticcheck -checks "${checks}" "${all_packages[@]}" 2>/dev/null || true)
+
+export IFS=$'\n'
+mapfile -t really_failing < <(sort -u <<<"${really_failing[*]}")
+unset IFS
+for pkg in "${failing_packages[@]}"; do
+  if ! kube::util::array_contains "$pkg" "${really_failing[@]}"; then
+    not_failing+=( "$pkg" )
+  fi
+done
+
+# Check that all failing_packages actually still exist
+gone=()
+for p in "${failing_packages[@]}"; do
+  if ! kube::util::array_contains "./$p" "${all_packages[@]}"; then
+    gone+=( "$p" )
+  fi
+done
+
+# Check to be sure all the packages that should pass check are.
+if [ ${#errors[@]} -eq 0 ]; then
+  echo 'Congratulations!  All Go source files have been checked.'
+else
+  {
+    echo "Errors from staticcheck:"
+    for err in "${errors[@]}"; do
+      echo "$err"
+    done
+    echo
+    echo 'Please review the above warnings. You can test via:'
+    echo "  staticcheck -checks \"${checks}\""
```

Comment:
nit: we should probably suggest using this script instead of staticcheck directly, since it also controls for the staticcheck version?